lac: Add AffineConstraints class

tamiko opened the pull request:
For now this is a verbatim copy of the ConstraintMatrix class. The idea
is to templatify this class and after that switch the code base to it.

My plan is:
 - [x] introduce a templated `AffineConstraints` object that is a copy of
       `ConstraintMatrix` but with a template parameter for the storage type
 - [x] switch the library to use `AffineConstraints`
 - [x] typedef `ConstraintMatrix` to `AffineConstraints<double>` (or a similar
       trick).

This has the advantage that we should get a good deal of backward
compatibility (for codebases that use `ConstraintMatrix`). And I need such an
intermediate stage for porting anyway. (I already tried simply templatifying
`ConstraintMatrix`, but *boy* that's messy.)

In reference to #1760, closes #1760

@dealii/dealii Quick question - is the name `AffineConstraints` OK (originally
suggested by @tjhei), or do you prefer something else. Further, discussion
about why we want this can be found at #1760 .

### Changelog
 - copied and renamed `ConstraintMatrix` to `AffineConstraints`
 - added a template parameter `template<typename number> AffineConstraints`
 - unified `number` (and similar) template parameters. All variants like `distribute_local_to_global` now only support uniform `number` types of matrices, vectors and the affine constraint object. (otherwise we would have ended up with 4 (FOUR) number type template parameters.)
 - instantiate `AffineConstraints` for `float`, `double`, `std::complex<float>`, and `std::complex<double>`.
 - refactored and unified all instantiations into `affine_constraints.inst.in`.
 - changed `create_mass_matrix` and `create_boundary_mass_matrix` to matchins matrix, vector and affine constraints number types.

tamiko commented:
OK - in order for this to not get out of hands (I am already templatifying for 3h). I will make the administrative decision that only matching vector / matrix and constraints number types are allowed. Anything else is silly anyway. (And that cleans up a lot of stuff.)

drwells commented:
forcing all number types to be the same might make things weird with complex solutions. Can we, e.g., set up hanging node constraints with `AffineConstraints<std::complex<double>>`?

tamiko commented:
@drwells 
>  Can we, e.g., set up hanging node constraints with AffineConstraints<std::complex<double>>?

Yes. The whole point of the exercise is that this will be possible.

The basic question is the following. Back when our vectors and matrices gained templates and we added a zoo of different vector wrappers, the `ConstraintMatrix` class itself never got updated. With that, we now have a zoo of templated functions like `distribute_local_to_global` etc, that in principle support all kinds of weird combinations of underlying number types.

But, is it realistic that you end up with something like
```
Vector<float> right_hand_side;
SparseMatrix<std::complex> system_matrix;
AffineConstraints<double> constraints;
```
Or, better, does it make sense to support such a use case? While I can see how upcasting from float to double to std::somplex is possible, I do not see how we should the converse. So, in order to simplify everything I am currently planning that only matching number types are allowed:
```
Vector<std::complex> right_hand_side;
SparseMatrix<std::complex> system_matrix;
AffineConstraints<std::complex> constraints;
```
The real question will be anyway how far I will come converting all usages of `ConstraintMatrix` with this...

tamiko commented:
On Thu, May 24, 2018, at 19:20 CDT, Denis Davydov <notifications@github.com> wrote:

> @tamiko 
>
>> only matching number types are allowed:
>
>>Vector<std::complex> right_hand_side;
>>SparseMatrix<std::complex> system_matrix;
>>AffineConstraints<std::complex> constraints;
>
> That sounds reasonable. The question is: do we have use cases where
> matrix elements in `AffineConstraints` are complex? I originally
> though that we templify only the inhomogeniety `b`. Probably @bangerth
> would know 😄

For periodic boundary conditions with phase-shift you need also
complex-valued coefficients (and that's something I am interested in).

As for the complexity, only adding a template parameter for the
inhomogeneity is already so much pain that I don't think we should omit
the entries. I am already working on this for 5h straigt...

drwells commented:
> But, is it realistic that you end up with something like

> `Vector<float> right_hand_side;`
> SparseMatrix<std::complex> system_matrix;`
> AffineConstraints<double> constraints;`

> Or, better, does it make sense to support such a use case?`

I'd prefer not to support this (I would hope obscure) use case, if for no other reason than to cut down on the massive number of template combinations we instantiate.

tamiko commented:
@drwells For reference. This is worse than the `dealii:Tensor` cleanup. Much worse...

drwells commented:
> @drwells For reference. This is worse than the dealii:Tensor cleanup. Much worse...

The Tensor cleanup is my favorite deal.II war story and I'm still impressed by the beautiful end results.

Given that our parallel (threads and MPI) static condensation algorithm is nearly sentient I suppose that I should not be surprised by your current problems.

If you can break things down into smaller pieces I would be willing to help. I suspect others would be willing to pitch in too!

One thing I have been meaning to do is refactor out `DataCache` into a new class `JaggedTable`. Would that, for example, help you right now?

tamiko commented:
> One thing I have been meaning to do is refactor out `DataCache` into a
> new class `JaggedTable`. Would that, for example, help you right now?

Right now it would help if no one would commit anything to
constraint_matrix.* :-D

It's going forward now. I am slowly starting to clean up the
instantiation mess and adding variants to the inst.in file.

tamiko commented:
So this compiles now. And I will happily go to bed now.

I will think about how to present the changes in this PR compared to `ConstraintMatrix` in a neat way. But, honestly, there was so much little fixing here and there, the diff compared to ConstraintMatrix will be a mess.

bangerth commented:
There is one use case I can think of with mixed real/complex arithmetic, and that is the projection of a complex-valued function onto a finite element space. The mass matrix you have to invert for that is still real-valued.

But I don't know how important that is. I'd say go for it the way you suggest right now. If we only templatize for the inhomogeneity, we know for a fact that 6 months from now someone will want to use complex-valued constraint coefficients as well.

tamiko commented:
Oh. Less than 10k line changes. I am actually suprised...

@bangerth @drwells Library compiles now without undefined references. I am currently on working out some minor bugs that I found while compiling the examples.

tamiko commented on include/deal.II/matrix_free/dof_info.h:
```diff
@@ -63,6 +63,7 @@ namespace internal
      * @author Katharina Kormann and Martin Kronbichler, 2010, 2011, 2018
      */
+    template <typename Number = double>
     struct DoFInfo
```
@kronbichler Heads up!

This is the most intrusive change that I made in order to support a templated `AffineConstraints` (aka `ConstraintMatrix`). I templatifyed `struct DoFInfo`. This change went remarkably smooth. There is only one minor caveat...

kronbichler replied:
This change goes a bit against the design of that class because it was designed to use the pure integer part, so a template at this place is confusing. Wouldn't it have been enough to simply put the template on the single function that passes in the AffineConstraints? On the other hand, I believe we should put most of that index functionality into DoFHandler anyway, as discussed in #2000, so we might accept this as a temporary hack. (Further assuming that we really do #2000 for the next release).

tamiko replied:
test results indicate that the testsuite isn't quite happy with this.

I will reroll the matrix free portion (without a template parameter) tomorrow and give this another try. Probably following your suggestion and only templatifying the function that passes in the constraints.

tamiko replied:
In any case - thanks a lot for the feedback :-)  (This kind of code review helps a lot. And to my defense - after working on the porting for 15h straight my brain was only half functional. So I didn't spot the obvious solution.)

tamiko commented on examples/step-37/step-37.cc:
```diff
@@ -871,7 +872,7 @@ namespace Step37
         DoFTools::extract_locally_relevant_level_dofs(
           dof_handler, level, relevant_dofs);
-        ConstraintMatrix level_constraints;
+        AffineConstraints<float> level_constraints;
```
... we now need a `AffineConstraints<float>` object for the level constraints (when you use a smoother operating on floating point numbers).

kronbichler replied:
This is also somewhat nasty, but we might get around this at a later time by allowing a different type of constraints as the `Number` type, as long as the argument here can be converted into `Number`.

tamiko commented on examples/step-59/step-59.cc:
```diff
@@ -997,8 +998,10 @@ namespace Step59
+    AffineConstraints<double> dummy_double;
+    dummy_double.close();
+    AffineConstraints<float> dummy_float;
+    dummy_float.close();
```
... This gets a bit messy in step-59 where we now need two dummy `AffineConstraints` objects.

tamiko commented on include/deal.II/matrix_free/cuda_matrix_free.h:
```diff
@@ -153,26 +153,27 @@ namespace CUDAWrappers
-           const AdditionalData    additional_data = AdditionalData());
+    reinit(const Mapping<dim> &             mapping,
+           const DoFHandler<dim> &          dof_handler,
+           const AffineConstraints<Number> &constraints,
```
@Rombur Can you please test whether these changes on cuda supported matrix free work as expected? I have no ability to test my code changes.

masterleinad replied:
For me, all the CUDA tests pass after fixing the compile errors.

tamiko commented on source/lac/tridiagonal_matrix.cc:
```diff
@@ -239,7 +241,8 @@ TridiagonalMatrix<number>::compute_eigenvalues()
   types::blas_int       info;
-  stev(&N, &nn, diagonal.data(), right.data(), nullptr, &one, nullptr, &info);
+  //   stev(&N, &nn, diagonal.data(), right.data(), nullptr, &one, nullptr,
+  //   &info);
```
@drwells @dealii/dealii This is an open TODO. We need a small internal wrapper that calls the LAPACK function `stev` for double and float, but is unimplemented otherwise

drwells replied:
We really need to clean up the LAPACK wrappers and find some tool that can generate these things automatically in the long run.

I am surprised that changing ConstraintMatrix resulted in breaks in this unrelated linear algebra class.

tamiko commented on include/deal.II/lac/affine_constraints.templates.h:
```diff
@@ -13,35 +13,1447 @@
+                    // of the list
+                    line->entries[entry] = std::make_pair(
+                      constrained_line->entries[0].first,
+                      constrained_line->entries[0].second * weight);
```
@dealii/dealii This is annoying. In case this gets instantiated for `std::complex<float>` this call to `make_pair` throws a warning with gcc (clang is fine):
```
Building CXX object source/lac/CMakeFiles/obj_lac_debug.dir/affine_constraints.cc.o
In file included from /usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include/g++-v7/bits/stl_algobase.h:64:0,
                 from /usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include/g++-v7/bits/stl_tree.h:63,
                 from /usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include/g++-v7/map:60,
                 from /usr/include/openmpi/ompi/mpi/cxx/mpicxx.h:40,
                 from /usr/include/x86_64-pc-linux-gnu/mpi.h:2703,
                 from /usr/include/mpi.h:10,
                 from /home/tamiko/workspace/dealii/build/include/deal.II/base/config.h:409,
                 from /home/tamiko/workspace/dealii/include/deal.II/lac/affine_constraints.h:19,
                 from /home/tamiko/workspace/dealii/source/lac/affine_constraints.cc:16:
/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include/g++-v7/bits/stl_pair.h: In instantiation of ‘constexpr std::pair<typename std::__decay_and_strip<_Tp>::__type, typename std::__decay_and_strip<_T2>::__type> std::make_pair(_T1&&, _T2&&) [with _T1 = const unsigned int&; _T2 = std::complex<float>; typename std::__decay_and_strip<_T2>::__type = std::complex<float>; typename std::__decay_and_strip<_Tp>::__type = unsigned int]’:
/home/tamiko/workspace/dealii/include/deal.II/lac/affine_constraints.templates.h:461:58:   required from ‘void dealii::AffineConstraints<number>::close() [with number = std::complex<float>]’
/home/tamiko/workspace/dealii/build/source/lac/affine_constraints.inst:22:16:   required from here
/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include/g++-v7/bits/stl_pair.h:519:5: note: the ABI of passing structure with complex float member has changed in GCC 4.4
     make_pair(_T1&& __x, _T2&& __y)
     ^~~~~~~~~
```

masterleinad replied:
Here, we could just use the explicit `std::pair<size_type, number>(...)`. That is even not much longer. When I try to compile this with `gcc-6.2.0` there is at least one more place which is not that obvious since it seems to be triggered by `boost`. Maybe, we should consider passing `-Wno-psabi `?

tamiko replied:
That's also what I thought. Except it didn't help.

masterleinad replied:
On `gcc-6.2.0` I can avoid the warning. Do you still see it?

tamiko commented on source/numerics/matrix_creator.inst.in:
```diff
@@ -67,42 +67,42 @@ for (scalar: REAL_AND_COMPLEX_SCALARS; deal_II_dimension : DIMENSIONS; deal_II_s
-     const Function<deal_II_space_dimension,numbers::NumberTraits<scalar>::real_type> * const coefficient,
-     const ConstraintMatrix   &constraints);
+     const Function<deal_II_space_dimension,scalar> * const coefficient,
+     const AffineConstraints<scalar>   &constraints);
```
@bangerth This is another intrusive change I made: I unified the interface for `create_mass_matrix` and `create_boundray_mass_matrix` to uniformly require the same number type. I think this is the right approach and the previous version that required complex-valued vectors but real-valued matrices is a cludge. First, we should really support mass matrices with a complex-valued coefficient because this shows up frequently in applications (think about dualized time-harmonic Maxwell's equation / complex-valued Helmholtz equation with a complex-valued permeability/permittiivity).

The only open problem with that is that ...

tamiko commented on include/deal.II/lac/solver_cg.h:
```diff
@@ -265,16 +269,17 @@ SolverCG<VectorType>::print_vectors(const unsigned int,
-  const std::vector<double> &diagonal,
-  const std::vector<double> &offdiagonal,
+  const std::vector<typename VectorType::value_type> &diagonal,
+  const std::vector<typename VectorType::value_type> &offdiagonal,
```
(above and below).

... we have to make `solver_cg` (that is used for one of the boundary variants) compatible with complex numbers. I have fixed compilation but this isn't tested for functionality yet and misses complex conjugation for the scalar product. But that should be an easy fix.

drwells commented:
I found the following failures: 
	2787 - dofs/dof_tools_02a.debug (Failed)
	2788 - dofs/dof_tools_02a.release (Failed)
	2795 - dofs/dof_tools_02b.debug (Failed)
	2796 - dofs/dof_tools_02b.release (Failed)
	2797 - dofs/dof_tools_02c.debug (Failed)
	2798 - dofs/dof_tools_02c.release (Failed)
	2799 - dofs/dof_tools_02d.debug (Failed)
	2800 - dofs/dof_tools_02d.release (Failed)
	2983 - dofs/sparsity_pattern.debug (Failed)
	2984 - dofs/sparsity_pattern.release (Failed)
	2985 - dofs/sparsity_pattern_01.debug (Failed)
	2986 - dofs/sparsity_pattern_01.release (Failed)
	3571 - fe/non_primitive_1.debug (Failed)
	3572 - fe/non_primitive_1.release (Failed)
	4717 - hp/create_laplace_matrix_02.debug (Failed)
	4718 - hp/create_laplace_matrix_02.release (Failed)
	4719 - hp/create_laplace_matrix_02b.debug (Failed)
	4720 - hp/create_laplace_matrix_02b.release (Failed)
	4725 - hp/create_laplace_matrix_04.debug (Failed)
	4726 - hp/create_laplace_matrix_04.release (Failed)
	4727 - hp/create_laplace_matrix_04b.debug (Failed)
	4728 - hp/create_laplace_matrix_04b.release (Failed)
	4737 - hp/create_mass_matrix_02.debug (Failed)
	4738 - hp/create_mass_matrix_02.release (Failed)
	4739 - hp/create_mass_matrix_02b.debug (Failed)
	4740 - hp/create_mass_matrix_02b.release (Failed)
	4745 - hp/create_mass_matrix_04.debug (Failed)
	4746 - hp/create_mass_matrix_04.release (Failed)
	4747 - hp/create_mass_matrix_04b.debug (Failed)
	4748 - hp/create_mass_matrix_04b.release (Failed)
	4749 - hp/create_mass_matrix_05.debug (Failed)
	4750 - hp/create_mass_matrix_05.release (Failed)
	5015 - hp/matrices.debug (Failed)
	5016 - hp/matrices.release (Failed)
	5017 - hp/matrices_hp.debug (Failed)
	5018 - hp/matrices_hp.release (Failed)
	5245 - lac/constraint_matrix_distribute_complex.debug (Failed)
	5246 - lac/constraint_matrix_distribute_complex.release (Failed)
	5331 - lac/gmres_eigenvalues.debug (Failed)
	5332 - lac/gmres_eigenvalues.release (Failed)
	5456 - lac/matrices_02.debug (Failed)
	5457 - lac/matrices_02.release (Failed)
	5500 - lac/precondition_chebyshev_01.debug (Failed)
	5501 - lac/precondition_chebyshev_01.release (Failed)
	5502 - lac/precondition_chebyshev_02.debug (Failed)
	5503 - lac/precondition_chebyshev_02.release (Failed)
	5504 - lac/precondition_chebyshev_03.debug (Failed)
	5505 - lac/precondition_chebyshev_03.release (Failed)
	5506 - lac/precondition_chebyshev_04.debug (Failed)
	5507 - lac/precondition_chebyshev_04.release (Failed)
	5508 - lac/precondition_chebyshev_05.debug (Failed)
	5509 - lac/precondition_chebyshev_05.release (Failed)
	5571 - lac/solver_signals.debug (Failed)
	5572 - lac/solver_signals.release (Failed)
	5623 - lac/utilities_01.debug (Failed)
	5624 - lac/utilities_01.release (Failed)
	5752 - lapack/solver_cg.debug (Failed)
	5753 - lapack/solver_cg.release (Failed)
	5754 - lapack/tridiagonal_ev1.debug (Failed)
	5755 - lapack/tridiagonal_ev1.release (Failed)
	6058 - matrix_free/compare_faces_by_cells.mpirun=1.debug (Failed)
	6059 - matrix_free/compare_faces_by_cells.mpirun=1.release (Failed)
	6060 - matrix_free/compare_faces_by_cells.mpirun=5.debug (Failed)
	6061 - matrix_free/compare_faces_by_cells.mpirun=5.release (Failed)
	6062 - matrix_free/compress_constraints.debug (Failed)
	6063 - matrix_free/compress_constraints.release (Failed)
	6066 - matrix_free/copy.debug (Failed)
	6067 - matrix_free/copy.release (Failed)
	6078 - matrix_free/estimate_condition_number_mass.debug (Failed)
	6079 - matrix_free/estimate_condition_number_mass.release (Failed)
	6112 - matrix_free/get_functions_float.debug (Failed)
	6113 - matrix_free/get_functions_float.release (Failed)
	6136 - matrix_free/integrate_functions_multife.debug (Failed)
	6137 - matrix_free/integrate_functions_multife.release (Failed)
	6138 - matrix_free/integrate_functions_multife2.debug (Failed)
	6139 - matrix_free/integrate_functions_multife2.release (Failed)
	6152 - matrix_free/inverse_mass_01.debug (Failed)
	6153 - matrix_free/inverse_mass_01.release (Failed)
	6232 - matrix_free/matrix_vector_15.debug (Failed)
	6233 - matrix_free/matrix_vector_15.release (Failed)
	6258 - matrix_free/matrix_vector_faces_02.debug (Failed)
	6259 - matrix_free/matrix_vector_faces_02.release (Failed)
	6308 - matrix_free/matrix_vector_float.debug (Failed)
	6309 - matrix_free/matrix_vector_float.release (Failed)
	6334 - matrix_free/multigrid_dg_periodic.mpirun=1.debug (Failed)
	6335 - matrix_free/multigrid_dg_periodic.mpirun=1.release (Failed)
	6336 - matrix_free/multigrid_dg_periodic.mpirun=3.debug (Failed)
	6337 - matrix_free/multigrid_dg_periodic.mpirun=3.release (Failed)
	6338 - matrix_free/multigrid_dg_periodic.mpirun=5.debug (Failed)
	6339 - matrix_free/multigrid_dg_periodic.mpirun=5.release (Failed)
	6340 - matrix_free/multigrid_dg_sip_01.mpirun=1.debug (Failed)
	6341 - matrix_free/multigrid_dg_sip_01.mpirun=1.release (Failed)
	6342 - matrix_free/multigrid_dg_sip_01.mpirun=5.debug (Failed)
	6343 - matrix_free/multigrid_dg_sip_01.mpirun=5.release (Failed)
	6344 - matrix_free/multigrid_dg_sip_02.mpirun=5.debug (Failed)
	6345 - matrix_free/multigrid_dg_sip_02.mpirun=5.release (Failed)
	6347 - matrix_free/parallel_multigrid.mpirun=3.debug (Failed)
	6348 - matrix_free/parallel_multigrid.mpirun=3.release (Failed)
	6349 - matrix_free/parallel_multigrid.mpirun=7.debug (Failed)
	6350 - matrix_free/parallel_multigrid.mpirun=7.release (Failed)
	6351 - matrix_free/parallel_multigrid_02.mpirun=3.debug (Failed)
	6352 - matrix_free/parallel_multigrid_02.mpirun=3.release (Failed)
	6353 - matrix_free/parallel_multigrid_02.mpirun=7.debug (Failed)
	6354 - matrix_free/parallel_multigrid_02.mpirun=7.release (Failed)
	6355 - matrix_free/parallel_multigrid_adaptive_01.mpirun=3.debug (Failed)
	6356 - matrix_free/parallel_multigrid_adaptive_01.mpirun=3.release (Failed)
	6357 - matrix_free/parallel_multigrid_adaptive_01.mpirun=7.debug (Failed)
	6358 - matrix_free/parallel_multigrid_adaptive_01.mpirun=7.release (Failed)
	6359 - matrix_free/parallel_multigrid_adaptive_02.mpirun=3.debug (Failed)
	6360 - matrix_free/parallel_multigrid_adaptive_02.mpirun=3.release (Failed)
	6361 - matrix_free/parallel_multigrid_adaptive_02.mpirun=7.debug (Failed)
	6362 - matrix_free/parallel_multigrid_adaptive_02.mpirun=7.release (Failed)
	6363 - matrix_free/parallel_multigrid_adaptive_03.mpirun=4.debug (Failed)
	6364 - matrix_free/parallel_multigrid_adaptive_03.mpirun=4.release (Failed)
	6365 - matrix_free/parallel_multigrid_adaptive_03.mpirun=7.debug (Failed)
	6366 - matrix_free/parallel_multigrid_adaptive_03.mpirun=7.release (Failed)
	6367 - matrix_free/parallel_multigrid_adaptive_04.mpirun=3.debug (Failed)
	6368 - matrix_free/parallel_multigrid_adaptive_04.mpirun=3.release (Failed)
	6369 - matrix_free/parallel_multigrid_adaptive_05.mpirun=4.debug (Failed)
	6370 - matrix_free/parallel_multigrid_adaptive_05.mpirun=4.release (Failed)
	6377 - matrix_free/parallel_multigrid_mf.mpirun=7.debug (Failed)
	6378 - matrix_free/parallel_multigrid_mf.mpirun=7.release (Failed)
	6383 - matrix_free/step-37.debug (Failed)
	6384 - matrix_free/step-37.release (Failed)
	6395 - matrix_free/stokes_computation.mpirun=4.debug (Failed)
	6396 - matrix_free/stokes_computation.mpirun=4.release (Failed)
	6397 - matrix_free/thread_correctness.debug (Failed)
	6398 - matrix_free/thread_correctness.release (Failed)
	6399 - matrix_free/thread_correctness_dg.debug (Failed)
	6400 - matrix_free/thread_correctness_dg.release (Failed)
	6401 - matrix_free/thread_correctness_hp.debug (Failed)
	6402 - matrix_free/thread_correctness_hp.release (Failed)
	6612 - mpi/condense_01.mpirun=2.debug (Failed)
	6613 - mpi/condense_01.mpirun=2.release (Failed)
	6614 - mpi/condense_01.mpirun=7.debug (Failed)
	6615 - mpi/condense_01.mpirun=7.release (Failed)
	6616 - mpi/constraint_matrix_condense_01.mpirun=1.debug (Failed)
	6617 - mpi/constraint_matrix_condense_01.mpirun=1.release (Failed)
	6618 - mpi/constraint_matrix_condense_01.mpirun=4.debug (Failed)
	6619 - mpi/constraint_matrix_condense_01.mpirun=4.release (Failed)
	7415 - mpi/step-37.mpirun=1.debug (Failed)
	7416 - mpi/step-37.mpirun=1.release (Failed)
	7417 - mpi/step-37.mpirun=7.debug (Failed)
	7418 - mpi/step-37.mpirun=7.release (Failed)
	7662 - multigrid/step-16-04.debug (Failed)
	7663 - multigrid/step-16-04.release (Failed)
	7664 - multigrid/step-16-05.debug (Failed)
	7665 - multigrid/step-16-05.release (Failed)
	7666 - multigrid/step-16-06.debug (Failed)
	7667 - multigrid/step-16-06.release (Failed)
	7805 - non_matching/coupling_01.debug (Failed)
	7806 - non_matching/coupling_01.release (Failed)
	7811 - non_matching/coupling_04.debug (Failed)
	7812 - non_matching/coupling_04.release (Failed)
	7813 - non_matching/coupling_05.mpirun=2.debug (Failed)
	7814 - non_matching/coupling_05.mpirun=2.release (Failed)
	7815 - non_matching/coupling_06.mpirun=2.debug (Failed)
	7816 - non_matching/coupling_06.mpirun=2.release (Failed)
	7833 - numerics/create_laplace_matrix_02.debug (Failed)
	7834 - numerics/create_laplace_matrix_02.release (Failed)
	7835 - numerics/create_laplace_matrix_02b.debug (Failed)
	7836 - numerics/create_laplace_matrix_02b.release (Failed)
	7841 - numerics/create_laplace_matrix_04.debug (Failed)
	7842 - numerics/create_laplace_matrix_04.release (Failed)
	7843 - numerics/create_laplace_matrix_04b.debug (Failed)
	7844 - numerics/create_laplace_matrix_04b.release (Failed)
	7859 - numerics/create_mass_matrix_02.debug (Failed)
	7860 - numerics/create_mass_matrix_02.release (Failed)
	7861 - numerics/create_mass_matrix_02b.debug (Failed)
	7862 - numerics/create_mass_matrix_02b.release (Failed)
	7867 - numerics/create_mass_matrix_04.debug (Failed)
	7868 - numerics/create_mass_matrix_04.release (Failed)
	7869 - numerics/create_mass_matrix_04b.debug (Failed)
	7870 - numerics/create_mass_matrix_04b.release (Failed)
	7871 - numerics/create_mass_matrix_05.debug (Failed)
	7872 - numerics/create_mass_matrix_05.release (Failed)

I did not run with scalapack installed so there might be more.

drwells commented:
Ah, now that the cdash results are up we don't need that giant comment.

I don't mind if you edit my comments (I would certainly not call that feature grotesque).

tamiko commented:
@drwells @masterleinad  Judging from the failing tests, my `SolverCG` hotfix seems to be totally botched and should be reverted and fixed properly. All instances of failing tests due to template deduction failure of `make_sparsity_pattern` should be resolved with the last commit.

If you two want to help out - feel free to commit to this branch. I suggest to not alter any tests/test outputs at this stage but to try to fix the incompatibilities :-)

kronbichler commented on include/deal.II/matrix_free/fe_evaluation.h:
```diff
@@ -2919,11 +2920,11 @@ inline FEEvaluationBase<dim, n_components_, Number, is_face>::FEEvaluationBase(
-         internal::MatrixFreeFunctions::DoFInfo::dof_access_face_exterior) :
-      internal::MatrixFreeFunctions::DoFInfo::dof_access_cell),
+    is_face ? (is_interior_face ? internal::MatrixFreeFunctions::DoFInfo<
+                                    Number>::dof_access_face_interior :
```
That is a strange place to break the line. Is this `clang-format`'s fault?

tamiko replied:
This is clang-format.

kronbichler commented on include/deal.II/matrix_free/fe_evaluation.h:
```diff
@@ -3413,9 +3415,9 @@ namespace internal
-    const internal::MatrixFreeFunctions::DoFInfo &dof_info)
+  check_vector_compatibility(const VectorType &vec,
+                             const internal::MatrixFreeFunctions::DoFInfo<
+                               typename VectorType::value_type> &dof_info)
```
Did you run all matrix-free tests already? This might break some functionality we once had, namely reading a `double` vector with an `FEEvaluation` object of type `float`. I have not checked whether we have a test (if not, I'm to blame), but this should be easy to fix in a subsequent PR.

kronbichler replied:
Ah we do, it is (at least) `matrix_free/step-37`.

tamiko replied:
@kronbichler Yes. : https://cdash.kyomu.43-1.org/viewTest.php?onlyfailed&buildid=2140

tamiko commented:
@drwells Let's not update the matrix-free tests at the moment. The interface changes that I made have to be reverted. (See @kronbichler 's explanation above)

drwells commented:
Ah, I missed that particular thread.

tamiko commented:
@masterleinad Please push the ABI change again. I am done fixing the indentation mess. :-/

tamiko commented:
@masterleinad I should have a fix for the matrix free stuff soon. I only see failing matrix_free tests on tester :-]

tamiko commented:
@kronbichler Please have a look at
https://github.com/dealii/dealii/pull/6664/commits/8b2dce3aa5348db7f4a4af18206743804b5f8f7c
and
https://github.com/dealii/dealii/pull/6664/commits/f252210fc73bc923f432ef08126df678e145e490

These are the two functional changes that I have to make. I am not overly happy with introducing an new template parameter but this is the only way how we can support `AffineConstraints<number>`. This passes all matrix-free tests (unaltered).

tamiko commented:
I want to see the world burn:
/run-tests

tamiko commented:
/run-tests

tamiko commented:
Down to 6 failing tests: Almost there! https://cdash.kyomu.43-1.org/viewTest.php?onlyfailed&buildid=2177

tamiko commented:
Down to 1 failing test. And that's a real gem: https://cdash.kyomu.43-1.org/testDetails.php?test=13461015&build=2177

tamiko commented:
@dealii/dealii All, this PR is now down to 0 failing tests. I only touched two tests (both in matrix_free). One trivial change (do not include `constraint_matrix.templates.h`) and one interface change in `matrix_free/step-37`.

With the current (more restrictive) interface, this test now requires an `AffineConstraints<float>` for the levels with float (instead of double). However, the current version of step-37 is fine without modification.
Ultimately, the question boils down to how generic the interfaces in `AffineConstraints` shall be - and my position is still that they should become more restrictive. And frankly, instantiating `distribute`, or `distribute_local_to_global` for every combination of `float`, `double`, `std::complex<float>`, and `std::complex<double>` for every type of input/output matrix, vector, or storage type in `AffineConstraints` is just excessive with very little gain.

I have stopped short of one goal I had. Namely, putting template constraints in place that ensure that number types are matching. The reason is very simple - we need mixed variants in matrix_free and multigrid at the moment. (I am also not overly sure of meshworker). That said, I will augment #6672 with more details for futue improvement. :-)

tamiko commented:
https://cdash.kyomu.43-1.org/index.php?project=deal.II

davydden commented:
> Ultimately, the question boils down to how generic the interfaces in AffineConstraints shall be - and my position is still that they should become more restrictive. And frankly, instantiating distribute, or distribute_local_to_global for every combination of float, double, std::complex<float>, and std::complex<double> for every type of input/output matrix, vector, or storage type in AffineConstraints is just excessive with very little gain.

I agree with that. 

I also checked the two matrix-free changes 

> @kronbichler Please have a look at 8b2dce3 and f252210

and they look reasonable to me, but let's see what @kronbichler have to say.

kronbichler commented:
> @kronbichler Please have a look at 8b2dce3 and f252210

I agree with these changes. In particular, the change in `step-37` only affects the matrix-based part that is there for reference (for the coarse solver) and leaves the mixed number types in vector access versus operator intact, so that is perfectly fine from a matrix-free point of view :wink:

Apart from the support for vector types on a `Number2` type different from the `Number` template in `FEEvaluation`, which is a feature we used to support, so there ought to be a backward-compatible way to keep it, I do definitely agree on trying to be narrow. We have lots of templates already and cannot handle yet another quadratic explosion.

masterleinad commented:
For `PETSc` with complex scalar type we still have failing tests:
```
	3519 - mpi/petsc_distribute_01.mpirun=1.debug (Failed)
	3520 - mpi/petsc_distribute_01.mpirun=10.debug (Failed)
	3521 - mpi/petsc_distribute_01.mpirun=2.debug (Failed)
	3522 - mpi/petsc_distribute_01.mpirun=4.debug (Failed)
	3523 - mpi/petsc_distribute_01_block.mpirun=1.debug (Failed)
	3524 - mpi/petsc_distribute_01_block.mpirun=10.debug (Failed)
	3525 - mpi/petsc_distribute_01_block.mpirun=2.debug (Failed)
	3526 - mpi/petsc_distribute_01_block.mpirun=4.debug (Failed)
	3527 - mpi/petsc_distribute_01_inhomogenous.mpirun=1.debug (Failed)
	3528 - mpi/petsc_distribute_01_inhomogenous.mpirun=10.debug (Failed)
	3529 - mpi/petsc_distribute_01_inhomogenous.mpirun=2.debug (Failed)
	3530 - mpi/petsc_distribute_01_inhomogenous.mpirun=4.debug (Failed)
	4005 - petsc_complex/assemble_01.mpirun=1.debug (Failed)
	4011 - petsc_complex/parallel_sparse_matrix_01.mpirun=3.debug (Failed)
```

masterleinad commented:
Down to one failing test after changing `ConstraintMatrix` to ` AffineConstraints<PetscScalar>`:
```
 4011: Test command: /export/home/darndt/.linuxbrew/Cellar/cmake/3.11.1/bin/cmake "-DTRGT=parallel_sparse_matrix_01.mpirun3.debug.diff" "-DTEST=petsc_complex/parallel_sparse_matrix_01.mpirun=3.debug" "-DEXPECT=PASSED" "-       DBINARY_DIR=/mnt/data/darndt/Sources/dealii-gcc-8/build/tests/petsc_complex" "-DGUARD_FILE=/mnt/data/darndt/Sources/dealii-gcc-8/build/tests/petsc_complex/parallel_sparse_matrix_01.debug/interrupt_guard.cc" "-P" "/mnt/data/   darndt/Sources/dealii-gcc-8/build/share/deal.II/scripts/run_test.cmake"
 4011: Test timeout computed to be: 600
 4011: /mnt/data/darndt/Sources/dealii-gcc-8/tests/petsc_complex/parallel_sparse_matrix_01.cc:90: error: undefined reference to 'void dealii::DoFTools::make_hanging_node_constraints<dealii::DoFHandler<2, 2>, std::              complex<double> >(dealii::DoFHandler<2, 2> const&, dealii::AffineConstraints<std::complex<double> >&)'
 4011: collect2: error: ld returned 1 exit status
 4011: make[3]: *** [parallel_sparse_matrix_01.debug/parallel_sparse_matrix_01.debug] Error 1
 4011: make[2]: *** [CMakeFiles/parallel_sparse_matrix_01.debug.dir/all] Error 2
 4011: make[1]: *** [CMakeFiles/parallel_sparse_matrix_01.mpirun3.debug.diff.dir/rule] Error 2
 4011: make: *** [parallel_sparse_matrix_01.mpirun3.debug.diff] Error 2
 4011: Test petsc_complex/parallel_sparse_matrix_01.mpirun=3.debug: BUILD
 4011: ===============================   OUTPUT BEGIN  ===============================
 4011: petsc_complex/parallel_sparse_matrix_01.mpirun=3.debug: BUILD failed. Output:
 4011: Scanning dependencies of target parallel_sparse_matrix_01.debug
 4011: [ 50%] Building CXX object CMakeFiles/parallel_sparse_matrix_01.debug.dir/parallel_sparse_matrix_01.cc.o
 4011: [ 50%] Building CXX object CMakeFiles/parallel_sparse_matrix_01.debug.dir/parallel_sparse_matrix_01.debug/interrupt_guard.cc.o
 4011: [ 50%] Linking CXX executable parallel_sparse_matrix_01.debug/parallel_sparse_matrix_01.debug
 4011: CMakeFiles/parallel_sparse_matrix_01.debug.dir/build.make:170: recipe for target 'parallel_sparse_matrix_01.debug/parallel_sparse_matrix_01.debug' failed
 4011: CMakeFiles/Makefile2:2115: recipe for target 'CMakeFiles/parallel_sparse_matrix_01.debug.dir/all' failed
 4011: CMakeFiles/Makefile2:2314: recipe for target 'CMakeFiles/parallel_sparse_matrix_01.mpirun3.debug.diff.dir/rule' failed
 4011: Makefile:1039: recipe for target 'parallel_sparse_matrix_01.mpirun3.debug.diff' failed
 4011:
 4011:
 4011: petsc_complex/parallel_sparse_matrix_01.mpirun=3.debug: ******    BUILD failed    *******
 4011:
 4011: ===============================    OUTPUT END   ===============================
 4011: Expected stage PASSED - aborting
 4011: CMake Error at /mnt/data/darndt/Sources/dealii-gcc-8/build/share/deal.II/scripts/run_test.cmake:140 (MESSAGE):
 4011:   *** abort
 4011:
 4011:
 1/1 Test #4011: petsc_complex/parallel_sparse_matrix_01.mpirun=3.debug ...***Failed   14.88 sec
 
 0% tests passed, 1 tests failed out of 1
 
 Total Test time (real) =  15.35 sec
 
 The following tests FAILED:
     4011 - petsc_complex/parallel_sparse_matrix_01.mpirun=3.debug (Failed)
 Errors while running CTest
```

masterleinad commented:
I am currently trying to fix `make_hanging_node_constraints` for `std::complex<double>`...

tamiko commented:
@masterleinad *umh* your last commit seems to have a lot of stray changes in it all over the place. Can you isolate the change you did for `PETScComplex`?

masterleinad commented:
> @masterleinad umh your last commit seems to have a lot of stray changes all over the place. Can you isolate the change you did for PETScComplex?

Sorry, strange things happened.

masterleinad commented:
For me, all tests are passing now.

tamiko commented:
@dealii/dealii All, I have regrouped, renamend and reapplied all changes. How shall we proceed with mergin/review? (And apropos testing - shall I run the full testsuite on that?)

davydden commented:
> shall I run the full testsuite on that

let's do that (with everything enabled) and also run complex PETSc/SLEPc as a second part.

drwells commented:
Nice work!

tamiko commented:
@drwells @masterleinad Thanks a lot for your help!

masterleinad commented:
> No failing tests for complex PETSc any more: https://cdash.kyomu.43-1.org/buildSummary.php?buildid=2236 :-)

The only caveat is that I had to replace `ConstraintMatrix` by `AffineConstraints<PetscSclar>` for complex `PETSc` scalar type.

tamiko commented:
@masterleinad Yes, but given the fact that PETSc with complex scalar type isn't a widely deployed mode I have little concern here. After all we do the whole exercise to put a matching complex number into the `AffineConstraints` object in the first place.

davydden commented:
@tamiko 

> When this PR is ready for review, I will split out minor parts that can be reviewed and merged quickly.

do you still want to do this? Or shall this PR be merged in its current form?

tamiko commented:
@davydden I changed my mind and would appreciate this PR to be merged in its current form. I will do the documentation update and more cleanup in a separate PR.

tamiko commented:
Thanks a lot. Your help is always much appreciated :smile:
Regarding the `-Wno-psabi` switch. I am not aware of any other way to stop gcc from warning loudly that gcc-4.4 changed the ABI for the following snippet of code:
```
#include <complex>
#include <algorithm>

int main()
{
  unsigned int a = 0.;
  std::complex<float> b = 0.;
  std::make_pair(a, b);
  return 0;
}
```
```
/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include/g++-v7/bits/stl_pair.h:519:5: note: the ABI of passing structure with complex float member has changed in GCC 4.4
     make_pair(_T1&& __x, _T2&& __y)
```

tamiko commented:
@dealii/dealii All, this was 30h of work. If we don't want to throw it away, we have to merge relatively soon. :smile_cat:

davydden commented:
> @dealii/dealii All, this was 30h of work. If we don't want to throw it away, we have to merge relatively soon. 😸

@dealii/dealii anyone objects merging this `AffineConstraints` PR? If not, @tamiko  please go ahead tomorrow.

davydden commented:
@tamiko do you recall changing any functionality in this PR? I am experience slower calculations (admittedly in debug mode) where some calculations (admittedly in Debug mode) became 3x slower.

tamiko commented:
@davydden The only functionality that was changed was that we added a template parameter to the class. I don't believe that some possible vtable lookups can amount to a slowdown of a factor of 3x.
Maybe something doesn't get inlined anymore with the change (or some similar indirect change).

Do you observe a measurable slowdown in release mode as well? (Are you sure it is exactly bisected to the introduction of AffineConstraints - or might it be something else that we changed in the meantime?)

@tjhei We really need the performance regression tests up again :-/

davydden commented:
> Are you sure it is exactly bisected to the introduction of AffineConstraints - or might it be something else that we changed in the meantime?

I am quite certain. I compare 2abb74f + cherry-picked missing instantiations on top of it: 757c128e8c   

vs 

b05181f which is one of the parents of the merge commit.

I will do more testing and edit this comment...

davydden commented:
@tamiko nevermind, it's all my mess up, see above.

tamiko commented:
@davydden No worries :-)
